### PR TITLE
exclude hidden directories instead of Removes recursion subdirectories

### PR DESCRIPTION
**Pull request by readerx**
The following scenarios are supported

There are a number of different types of configurations, like:
/var/jenkins_home/casc_configs/kubernetes-slave/<slave1.yamls、slave2.yaml .......>
/var/jenkins_home/casc_configs/credentials/<credentials1.yaml、credentials2.yaml.......>
/var/jenkins_home/casc_configs/default/<a.yaml、b.yaml........>
/var/jenkins_home/casc_configs/authorization-strategy/<auth1.yaml、auth2.yaml.......>

#629 
#751

### CODE REVIEW

**Comment by casz** [comment]
Could you please revert the formatting? makes the PR harder to review @readerx

**Comment by readerx** [comment]
> 
> 
> Could you please revert the formatting? makes the PR harder to review @readerx

I have revert the formatting
Use the smallest possible changes to complete the function. 
Any other things that need to be modified?

@casz

**Comment by readerx** [comment]
> 
> 
> The code LGTM, perhaps someone with a k8s cluster can confirm it works as desired?

I have already confirmed it on the my k8s cluster.

**Comment by casz** [comment]
@readerx good enough for me

@ndeloof would you mind taking a look?

**Comment by olivergondza** [comment]
The question whether to include this or not was coined on office hours today. There seems to be use-cases for both of the behaviors so the idea coined there was using ant globs we use on so many places in Jenkins so customers have better control of what to include:

- `/config/drive/jenkins.yaml`
- `/config/drive/*.yaml`
- `/config/drive/*/*.yaml`
- `/config/drive/**/*.yaml`

The downside I see is when this gets incorrectly quoted, the globs can be interpreted by shell before Jenkins is launched causing surprising results.

**Comment by readerx** [comment]
> 
> 
> The question whether to include this or not was coined on office hours today. There seems to be use-cases for both of the behaviors so the idea coined there was using ant globs we use on so many places in Jenkins so customers have better control of what to include:
> 
>     * `/config/drive/jenkins.yaml`
> 
>     * `/config/drive/*.yaml`
> 
>     * `/config/drive/*/*.yaml`
> 
>     * `/config/drive/**/*.yaml`
> 
> 
> The downside I see is when this gets incorrectly quoted, the globs can be interpreted by shell before Jenkins is launched causing surprising results.

This issue can be discussed separately

**Comment by readerx** [comment]
Could you please review this PR?    @ndeloof 

1. Traversing the directory, this is a use case.
2. Can't remove this function because of k8s configmap.（Although k8s uses ../... to perform directory mapping, we can solve the problem by not traversing the hidden directory.）
3. Excluding hidden directories should be appropriate, as hidden directories will not be used under normal.

If there are other good ways I can do it.

File: plugin/src/main/java/io/jenkins/plugins/casc/ConfigurationAsCode.java [hunk 1]
```diff
@@ -569,20 +569,25 @@ private void configureWith(List<YamlSource> sources) throws ConfiguratorExceptio
...
-                    .collect(toList());
+        try (Stream<Path> stream = Files.find(Paths.get(path), Integer.MAX_VALUE,
+                (next, attrs) -> !attrs.isDirectory() && !isHidden(next) && matcher.matches(next))) {
+            return stream.collect(toList());
         } catch (IOException e) {
             throw new IllegalStateException("failed config scan for " + path, e);
         }
     }
 
+    private static boolean isHidden(Path path) {
```
Comment (casz) [inline plugin/src/main/java/io/jenkins/plugins/casc/ConfigurationAsCode.java]: Would it not be better to use `File.isHidden`?

Comment (readerx) [reply plugin/src/main/java/io/jenkins/plugins/casc/ConfigurationAsCode.java]: E.g：
/var/jenkins_home/casc_configs/..data/abc.yaml 

It is not needed， but File.isHidden can't handle

Comment (casz) [reply plugin/src/main/java/io/jenkins/plugins/casc/ConfigurationAsCode.java]: Fair enough 😄

Comment (readerx) [reply plugin/src/main/java/io/jenkins/plugins/casc/ConfigurationAsCode.java]: Is it possible to merge into the master? 😀

**Comment by ewelinawilkosz** [comment]
sorry @casz, I requested review from you accidentally

@readerx would you consider updating documentation? README.md contains a section about `CASC_JENKINS_CONFIG` where you could elaborate on new behavior

**Comment by readerx** [comment]
> 
> 
> sorry @casz, I requested review from you accidentally
> 
> @readerx would you consider updating documentation? README.md contains a section about `CASC_JENKINS_CONFIG` where you could elaborate on new behavior

I am working hard on my English.
But I am willing to give it a try 😁

**Comment by readerx** [comment]
I have modified the README.md. @ewelinawilkosz

File: README.md [hunk 2]
```diff
@@ -55,6 +55,8 @@ Second, the plugin looks for the `CASC_JENKINS_CONFIG` environment variable. The
 - A full path to a single file. For example, `/var/jenkins_home/casc_configs/jenkins.yaml`.
 - A URL pointing to a file served on the web. For example, `https://acme.org/jenkins.yaml`.
 
+If `CASC_JENKINS_CONFIG` point to a folder, the plugin will recursively traversing the folder to find file(suffix with .yml,.yaml,.YAML,.YML), but dosn't contain hidden files or hidden subdirectories. It dosn't follow symbolic links.
```
Comment (ewelinawilkosz) [inline README.md]: couple of typos:
"point**s**", "travers**e**", space between "file" and "(", and "do**e**sn't" and we're good to go!

Comment (readerx) [reply README.md]: thank you.
I have modified it according to your statement.

**Comment by DragonBay** [comment]
why revert this commit? it is a very useful feature

**Comment by ewelinawilkosz** [comment]
the revert was clicked by mistake (I didn't even notice...), there is no plan to revert it

**Pull request merged**